Sources/NIO/ByteBuffer-aux.swift

```diff
@@ -431,4 +431,178 @@ extension ByteBuffer {
             return $0
         }
     }
+
+    @discardableResult
+    public mutating func writeImmutableBuffer(_ buffer: ByteBuffer) -> Int {
+        var mutable = buffer
+        return self.writeBuffer(&mutable)
+    }
+}
+
+extension ByteBuffer {
+    /// Return an empty `ByteBuffer` allocated with `ByteBufferAllocator()`.
+    ///
+    /// Calling this constructor will not allocate because it will return a `ByteBuffer` that wraps a shared storage
+    /// object. As soon as you write to the constructed buffer however, you will incur an allocation because a
+    /// copy-on-write will happen.
+    @inlinable
+    public init() {
+        self = ByteBufferAllocator.zeroCapacityWithDefaultAllocator
+    }
+
+    /// Create a fresh `ByteBuffer` containing the bytes of the `string` encoded as UTF-8.
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space fitting `string` and potentially some extra space using
+    /// Swift's default allocator.
+    ///
+    /// - info: If you have access to a `Channel`, `ChannelHandlerContext`, or `ByteBufferAllocator` it is
+    ///         recommended using `channel.allocator.buffer(capacity: ...)` to allocate a `ByteBuffer` of the right
+    ///         size followed by a `writeString` instead of using this method. This allows SwiftNIO to do accounting
+    ///         and optimisations of resources acquired for operations on a given `Channel` in the future.
+    public init(string: String) {
+        self = ByteBufferAllocator().buffer(capacity: string.utf8.count)
+        self.writeString(string)
+    }
+
+    /// Create a fresh `ByteBuffer` containing the bytes of the `string` encoded as UTF-8.
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space fitting `string` and potentially some extra space using
+    /// Swift's default allocator.
+    ///
+    /// - info: If you have access to a `Channel`, `ChannelHandlerContext`, or `ByteBufferAllocator` it is
+    ///         recommended using `channel.allocator.buffer(capacity: ...)` to allocate a `ByteBuffer` of the right
+    ///         size followed by a `writeSubstring` instead of using this method. This allows SwiftNIO to do accounting
+    ///         and optimisations of resources acquired for operations on a given `Channel` in the future.
+    public init(substring string: Substring) {
```

Comment:
Is it too slow to provide a generic StringProtocol initialiser?

Reply:
used to be too slow, now it's API and we should stay consistent.